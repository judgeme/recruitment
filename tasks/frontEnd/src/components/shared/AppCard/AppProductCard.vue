<script setup lang="ts">
import { defineProps, defineOptions } from 'vue';

import AppIcon from '@/components/shared/AppIcon/index.vue';
import AppLink from '@/components/shared/AppLink/index.vue';

type Props = {
  productUrl: string;
  shopUrl: string;
  imageUrl?: string;
  productName: string;
  price: string | number;
  extras?: {
    status: string;
    discount: string;
  };
};

defineProps<Props>();
// set display name in dev-tools
defineOptions({
  name: 'AppProductCard'
});
</script>
<template>
  <div class="bg-white relative w-full overflow-hidden group">
    <!-- media section -->
    <app-link
      :to="productUrl"
      class="w-full block h-[380px] relative overflow-hidden"
      :title="productName"
    >
      <figure class="w-full h-full transition-all group-hover:opacity-70">
        <img :src="imageUrl" class="w-full h-full block" :alt="productName" decoding="async" />
      </figure>
      <div v-if="extras" class="absolute top-0 left-0 right-0 py-4 px-6 flex justify-between">
        <span class="text-primary text-tiny leading-6 tracking-[1px] font-medium">{{
          extras.status
        }}</span>
        <span class="text-primary text-tiny leading-6 tracking-[1px] font-medium">{{
          extras.discount
        }}</span>
      </div>
      <div
        class="absolute bottom-0 left-0 right-0 w-full px-4 flex justify-center items-center h-7 transition-all md:translate-y-full md:group-hover:translate-y-0"
      >
        <span
          class="inline-flex items-center bg-black text-white py-0.5 px-4 text-small font-bold h-full uppercase tracking-mono"
        >
          <span>Quick Look</span>
        </span>
        <span
          class="relative inline-flex align-middle m-0 py-0.5 px-2 bg-grey-deep justify-center items-center h-full text-center text-white"
        >
          <app-icon icon="eye" :size="16" viewBox="0 0 512 512" />
        </span>
      </div>
    </app-link>
    <!-- info section -->
    <div class="relative w-full mt-6 px-6 text-center">
      <p class="uppercase text-sm tracking-double text-primary mb-2">{{ productName }}</p>
      <div
        class="text-grey-80 text-sm transition-all md:opacity-100 md:group-hover:opacity-0 md:group-hover:translate-x-15"
      >
        <bdi>
          <span>$</span>
          {{ price }}
        </bdi>
      </div>
      <div
        class="font-semibold md:absolute md:bottom-0 md:left-1/2 transition-all md:opacity-0 md:group-hover:opacity-100 md:group-hover:-translate-x-1/2 md:-translate-x-full"
      >
        <app-link :to="shopUrl" class="text-xs uppercase text-grey-80 transition-colors"
          >Visit Store</app-link
        >
      </div>
    </div>
  </div>
</template>
