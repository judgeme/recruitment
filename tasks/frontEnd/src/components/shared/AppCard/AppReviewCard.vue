<script setup lang="ts">
import { defineProps, withDefaults, defineOptions, computed } from 'vue';
import AppRating from '@/components/shared/AppRating/index.vue';
import AppIcon from '@/components/shared/AppIcon/index.vue';
import type { ReviewModel } from '@/models';

// set display name in dev-tools
defineOptions({
  name: 'AppReviewCard'
});

const emit = defineEmits<{
  openImage: [value: string];
  onEditReview: [value: ReviewModel];
  onDeleteReview: [value: string];
}>();

const props = withDefaults(defineProps<ReviewModel>(), {});

/*notice api ratings are greater than 5, 
and they are strings instead of numbers
maybe not be needed in other scenario */
const rating = computed(() => {
  const value = props.review.rating;
  const newValue = typeof value === 'string' ? parseInt(value) : value;
  if (newValue > 5 && newValue < 10) {
    return newValue / 2;
  }
  return newValue;
});

const onOpenImage = (img: string) => {
  emit('openImage', img);
};
const onEditReview = () => {
  // emit the current review object
  emit('onEditReview', props);
};
const onDeleteReview = (id: string) => {
  emit('onDeleteReview', id);
};
</script>
<template>
  <div
    class="flex flex-col md:flex-row justify-start gap-10 border-grey-60 last-of-type:border-0 last-of-type:pb-0 border-b py-6"
  >
    <div class="w-full md:w-60">
      <p class="text-sm text-primary font-medium mb-2 truncate" :title="reviewer.name">
        {{ reviewer.name }}
      </p>
      <p class="text-xs font-light text-primary truncate mb-4" :title="reviewer.email">
        {{ reviewer.email }}
      </p>
      <p class="text-xs font-light text-primary truncate mb-4" v-if="review.verifiedPurchase">
        Verified Purchase ✔️
      </p>
      <span class="text-xs font-light text-white truncate p-2 bg-primary" :title="reviewer.email">
        {{ shopName }}
      </span>
      <div class="flex gap-5 mt-10">
        <button
          class="p-2 text-danger transition-colors rounded-full bg-white shadow-lg hover:bg-primary"
          @click="onDeleteReview(id)"
        >
          <app-icon icon="trash" :size="20" />
        </button>
        <button
          class="p-2 text-primary transition-colors rounded-full bg-white shadow-lg hover:bg-primary hover:text-white"
          @click="onEditReview"
        >
          <app-icon icon="pen" :size="20" />
        </button>
      </div>
    </div>
    <!-- rating and body -->
    <div class="w-100 max-w-100">
      <!-- stars -->
      <div class="mb-4">
        <app-rating :model-value="rating" disabled />
      </div>
      <p class="text-sm text-primary font-medium mb-2">{{ review.reviewTitle }}</p>
      <div class="text-xs font-light text-primary">{{ review.reviewBody }}</div>
    </div>
    <!-- photos -->
    <div class="flex-1">
      <ul class="flex flex-wrap gap-3">
        <li
          v-for="reviewImage in review.reviewImages"
          :key="reviewImage"
          class="flex basis-full md:basis-1/4 items-center w-25 h-25"
        >
          <button
            class="bg-grey-60 h-full w-full border-grey-80 overflow-hidden"
            @click="onOpenImage(reviewImage)"
          >
            <img class="w-full h-full" :src="reviewImage" :alt="reviewImage" decoding="async" />
          </button>
        </li>
      </ul>
    </div>
  </div>
</template>
