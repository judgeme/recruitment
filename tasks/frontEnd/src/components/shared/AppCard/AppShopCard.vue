<script setup lang="ts">
import { defineProps, defineOptions } from 'vue';

import AppLink from '@/components/shared/AppLink/index.vue';

type Props = {
  shopUrl: string;
  imageUrl?: string;
  shopName: string;
  location?: string;
};

defineProps<Props>();
// set display name in dev-tools
defineOptions({
  name: 'AppShopCard'
});
</script>
<template>
  <div class="bg-white relative w-full overflow-hidden group">
    <!-- media section -->
    <app-link
      :to="shopUrl"
      class="w-full block h-50 relative overflow-hidden"
      :title="shopName"
    >
      <figure class="w-full h-full transition-all group-hover:opacity-70">
        <img :src="imageUrl" class="w-full h-full block" :alt="shopName" decoding="async" />
      </figure>
    </app-link>
    <!-- info section -->
    <div class="relative w-full mt-6 px-6 text-center">
      <p class="uppercase text-sm tracking-double text-primary mb-2">{{ shopName }}</p>
      <p class="uppercase text-xs text-primary mb-2">({{ location }})</p>
      <div class="font-semibold transition-all">
        <app-link :to="shopUrl" class="text-xs uppercase text-grey-80 transition-colors"
          >Visit Store</app-link
        >
      </div>
    </div>
  </div>
</template>
