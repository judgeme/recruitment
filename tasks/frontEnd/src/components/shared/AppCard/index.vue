<script setup lang="ts">
import { defineOptions } from 'vue';

defineOptions({
  name: 'AppCard'
});
</script>
<template>
  <div class="w-full block bg-white p-6">
    <slot />
  </div>
</template>