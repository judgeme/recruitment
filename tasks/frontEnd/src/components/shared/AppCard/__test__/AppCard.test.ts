import { mount } from '@vue/test-utils';
import { describe, it, expect } from 'vitest';
import AppCard from '@/components/shared/AppCard/index.vue'; 

describe('AppCard.vue', () => {
  it('renders card with provided slot content', () => {
    const slotContent = 'This is the content of the card';

    const wrapper = mount(AppCard, {
      slots: {
        default: slotContent,
      },
    });

    // Check if the card element is rendered
    const card = wrapper.find('div');
    expect(card.exists()).toBe(true);

    // Check if the provided slot content is rendered
    expect(card.text()).toContain(slotContent);

    expect(wrapper.html()).toMatchSnapshot();
  });
});
