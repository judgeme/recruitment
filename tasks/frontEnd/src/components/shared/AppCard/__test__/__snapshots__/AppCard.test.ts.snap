// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppCard.vue > renders card with provided slot content 1`] = `"<div class="w-full block bg-white p-6">This is the content of the card</div>"`;
